### PR TITLE
Add experimental Bazel support

### PR DESCRIPTION
**Pull request by jin**
This adds build and test support with Bazel using the `android_instrumentation_test` rule. This is experimental as it requires Bazel built from master, and there are several known issues as listed in the README. 

Since Bazel does not generate the AndroidManfiest.xml automatically for the test application like Gradle does, they have been created manually.

The README section also recommends Bazel-related issues to be filed against the Bazel repository.

### CODE REVIEW

**Comment by jin** [comment]
Note that the README steps are missing the test arg `--install_test_services=true`. I have an internal change that adds that to the test stub script.

**Comment by jin** [comment]
@slinzner, PTAL?

File: ui/espresso/BasicSample/BUILD.bazel [hunk 1]
```diff
@@ -21,7 +20,7 @@ android_library(
 android_binary(
```
Comment (nkoroste) [inline ui/espresso/BasicSample/BUILD.bazel]: Off topic, can you add license/copyright to these bazel files

File: ui/espresso/BasicSample/app/src/main/EmptyManifest.xml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (nkoroste) [inline ui/espresso/BasicSample/app/src/main/EmptyManifest.xml]: The file name says empty but it's no longer empty

Comment (nkoroste) [reply ui/espresso/BasicSample/app/src/main/EmptyManifest.xml]: Same for the rest of the manifest files

Comment (jin) [reply ui/espresso/BasicSample/app/src/main/EmptyManifest.xml]: Done, renamed to AppManifest.xml

File: ui/espresso/BasicSample/BUILD.bazel (hunk 1 above)
Comment (jin) [reply ui/espresso/BasicSample/BUILD.bazel]: Done, added license header.

**Comment by jin** [comment]
Changed `target_device` to use `generic_phone` instead of `nexus_6`.

File: ui/espresso/DataAdapterSample/app/src/main/AppManifest.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (C) 2015 The Android Open Source Project
```
Comment (slinzner) [inline ui/espresso/DataAdapterSample/app/src/main/AppManifest.xml]: 2018

File: ui/espresso/IdlingResourceSample/app/src/main/AppManifest.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright 2016, The Android Open Source Project
```
Comment (slinzner) [inline ui/espresso/IdlingResourceSample/app/src/main/AppManifest.xml]: 2018

File: ui/espresso/IntentsAdvancedSample/app/src/main/AppManifest.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+~ Copyright (C) 2015 The Android Open Source Project
```
Comment (slinzner) [inline ui/espresso/IntentsAdvancedSample/app/src/main/AppManifest.xml]: 2018

File: ui/espresso/IntentsBasicSample/app/src/main/AppManifest.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+~ Copyright (C) 2015 The Android Open Source Project
```
Comment (slinzner) [inline ui/espresso/IntentsBasicSample/app/src/main/AppManifest.xml]: 2018

File: ui/espresso/DataAdapterSample/app/src/main/AppManifest.xml (hunk 3 above)
Comment (jin) [reply ui/espresso/DataAdapterSample/app/src/main/AppManifest.xml]: Done.

File: ui/espresso/IdlingResourceSample/app/src/main/AppManifest.xml (hunk 4 above)
Comment (jin) [reply ui/espresso/IdlingResourceSample/app/src/main/AppManifest.xml]: Done.

File: ui/espresso/IntentsAdvancedSample/app/src/main/AppManifest.xml (hunk 5 above)
Comment (jin) [reply ui/espresso/IntentsAdvancedSample/app/src/main/AppManifest.xml]: Done.

File: ui/espresso/IntentsBasicSample/app/src/main/AppManifest.xml (hunk 6 above)
Comment (jin) [reply ui/espresso/IntentsBasicSample/app/src/main/AppManifest.xml]: Done.

**Comment by jin** [comment]
Ran ./test_all.sh

```
// pruned

ui/uiautomator/BasicSample :app:transformNativeLibsWithStripDebugSymbolForDebug UP-TO-DATE
unit/BasicSample :app:transformDexArchiveWithExternalLibsDexMergerForDebug UP-TO-DATE
unit/BasicSample :app:transformDexArchiveWithDexMergerForDebug UP-TO-DATE
unit/BasicSample :app:mergeDebugJniLibFolders UP-TO-DATE
unit/BasicSample :app:transformNativeLibsWithMergeJniLibsForDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformDexArchiveWithDexMergerForDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:mergeDebugJniLibFolders UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformNativeLibsWithStripDebugSymbolForDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformResourcesWithMergeJavaResForDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:validateSigningDebug
unit/BasicUnitAndroidTest :app:packageDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:assembleDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:bundleAppClassesDebug UP-TO-DATE
unit/BasicUnitAndroidTest :app:preDebugAndroidTestBuild UP-TO-DATE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestAidl UP-TO-DATE
unit/BasicUnitAndroidTest :app:processDebugAndroidTestManifest UP-TO-DATE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestRenderscript UP-TO-DATE
unit/BasicUnitAndroidTest :app:generateDebugAndroidTestBuildConfig UP-TO-DATE
unit/BasicUnitAndroidTest :app:generateDebugAndroidTestResValues UP-TO-DATE
unit/BasicUnitAndroidTest :app:generateDebugAndroidTestResources UP-TO-DATE
unit/BasicUnitAndroidTest :app:mergeDebugAndroidTestResources UP-TO-DATE
unit/BasicUnitAndroidTest :app:splitsDiscoveryTaskDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:processDebugAndroidTestResources UP-TO-DATE
unit/BasicUnitAndroidTest :app:generateDebugAndroidTestSources UP-TO-DATE
unit/BasicUnitAndroidTest :app:javaPreCompileDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestJavaWithJavac UP-TO-DATE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestNdk NO-SOURCE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestSources UP-TO-DATE
unit/BasicUnitAndroidTest :app:mergeDebugAndroidTestShaders UP-TO-DATE
unit/BasicUnitAndroidTest :app:compileDebugAndroidTestShaders UP-TO-DATE
unit/BasicUnitAndroidTest :app:generateDebugAndroidTestAssets UP-TO-DATE
unit/BasicUnitAndroidTest :app:mergeDebugAndroidTestAssets UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformClassesWithDexBuilderForDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformDexArchiveWithExternalLibsDexMergerForDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformDexArchiveWithDexMergerForDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:mergeDebugAndroidTestJniLibFolders UP-TO-DATE
unit/BasicUnitAndroidTest :app:transformNativeLibsWithMergeJniLibsForDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:processDebugAndroidTestJavaRes NO-SOURCE
unit/BasicUnitAndroidTest :app:transformResourcesWithMergeJavaResForDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:validateSigningDebugAndroidTest
unit/BasicUnitAndroidTest :app:packageDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:assembleDebugAndroidTest UP-TO-DATE
unit/BasicUnitAndroidTest :app:connectedDebugAndroidTest
unit/BasicUnitAndroidTest Starting 1 tests on emulator-5554 - 8.0.0
unit/BasicUnitAndroidTest :app:connectedAndroidTest
unit/BasicUnitAndroidTest
unit/BasicUnitAndroidTest BUILD SUCCESSFUL in 2s
unit/BasicUnitAndroidTest 66 actionable tasks: 3 executed, 63 up-to-date

ALL TESTS PASS
```

File: ui/espresso/BasicSample/app/src/main/java/com/example/android/testing/espresso/BasicSample/MainActivity.java [hunk 7]
```diff
@@ -54,16 +54,17 @@ public void onClick(View view) {
...
-            case R.id.changeTextBt:
-                // First button's interaction: set a text in a text view.
-                mTextView.setText(text);
-                break;
-            case R.id.activityChangeTextBtn:
-                // Second button's interaction: start an activity and send a message to it.
-                Intent intent = ShowTextActivity.newStartIntent(this, text);
-                startActivity(intent);
-                break;
+        final int changeTextBtId = R.id.changeTextBt;
```
Comment (JoseAlcerreca) [inline ui/espresso/BasicSample/app/src/main/java/com/example/android/testing/espresso/BasicSample/MainActivity.java]: Unrelated change?

File: ui/uiautomator/BasicSample/app/src/main/java/com/example/android/testing/uiautomator/BasicSample/MainActivity.java [hunk 8]
```diff
@@ -54,16 +54,16 @@ public void onClick(View view) {
...
-            case R.id.changeTextBt:
-                // First button's interaction: set a text in a text view.
-                mTextView.setText(text);
-                break;
-            case R.id.activityChangeTextBtn:
-                // Second button's interaction: start an activity and send a message to it.
-                Intent intent = ShowTextActivity.newStartIntent(this, text);
-                startActivity(intent);
-                break;
+        final int changeTextBtId = R.id.changeTextBt;
```
Comment (JoseAlcerreca) [inline ui/uiautomator/BasicSample/app/src/main/java/com/example/android/testing/uiautomator/BasicSample/MainActivity.java]: Unrelated change?

File: ui/espresso/BasicSample/app/src/main/java/com/example/android/testing/espresso/BasicSample/MainActivity.java (hunk 7 above)
Comment (jin) [reply ui/espresso/BasicSample/app/src/main/java/com/example/android/testing/espresso/BasicSample/MainActivity.java]: This is required. Bazel doesn't finalize R ids at the `android_library` level in order for `aapt` to process/merge the resources at the `android_binary` level. [This is a good explanation](https://blog.bazel.build/2018/02/14/how-android-builds-work-in-bazel.html) of why Bazel needs this:

> The resource ID values generated for android_library targets are only temporary, since higher-level targets might depend on multiple targets where different resources were assigned the same ID. To ensure that resource IDs aren’t persisted anywhere permanent, the R files record the IDs as nonfinal, ensuring that compilation doesn’t inline them into other Java code. Additionally, an android_library's R files should be discarded after building is complete.

File: ui/uiautomator/BasicSample/app/src/main/java/com/example/android/testing/uiautomator/BasicSample/MainActivity.java (hunk 8 above)
Comment (jin) [reply ui/uiautomator/BasicSample/app/src/main/java/com/example/android/testing/uiautomator/BasicSample/MainActivity.java]: Same as above.
